[notifications] remove .native hardcoded platform imports

marklawlor opened the pull request:
# Why

Fix: #26248

# How

Multiple imports have been hardcoded to import the `.native` file, circumventing platform extensions. This causes the Web to try and use native APIs. 

To appease TypeScript, I renamed the `.web.ts` files to simply `.ts`. I also adjusted the web types to match the `.native` modules, as they were using `null` instead of `boolean`. The only exception to this is commented below

# Test Plan

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

marklawlor committed:
[notifications] remove .native hardcoded platform imports

marklawlor committed:
Update packages/expo-notifications/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

marklawlor commented on packages/expo-notifications/src/setNotificationCategoryAsync.ts:
```diff
@@ -23,7 +23,7 @@ export default async function setNotificationCategoryAsync(
   actions: NotificationAction[],
   options?: NotificationCategoryOptions
-): Promise<NotificationCategory> {
+): Promise<NotificationCategory | null> {
```
This is a public facing type change. I'm unsure of the consequences of it.

wschurman replied:
Interesting. I think it's probably fine but it could definitely cause a `tsc` error in people's apps (which is okay, we'll just want to make sure it's in the changelog).

tsapeta replied:
Yeah, it seems fine for this simple change, I wouldn't consider it as breaking 👍

marklawlor committed:
remove stale build files

wschurman commented on packages/expo-notifications/src/NotificationCategoriesModule.ts:
```diff
@@ -1,8 +1,8 @@
 export default {
-  async getNotificationCategoriesAsync(): Promise<null> {
-    return null;
+  async getNotificationCategoriesAsync(): Promise<never[]> {
```
```suggestion
  async getNotificationCategoriesAsync(): Promise<NotificationCategory[]> {
```

(so that the types of the default import/export are correct in apps)

wschurman commented on packages/expo-notifications/src/NotificationCategoriesModule.ts:
```diff
@@ -11,7 +11,7 @@ export default {
   ): Promise<null> {
```
```suggestion
  ): Promise<NotificationCategory | null> {
```

(also make change to `NotificationCategoriesModule.types.ts`)

wschurman replied:
Also add the correct types to `options` arg for this since this is now the file used to dictate the types of the methods. Can also use `NotificationCategoryOptions` in `NotificationCategoriesModule.types.ts`
```typescript
options?: NotificationCategoryOptions
```

marklawlor committed:
Update packages/expo-notifications/src/NotificationCategoriesModule.ts

Co-authored-by: Will Schurman <wschurman@expo.io>

marklawlor committed:
Update packages/expo-notifications/src/NotificationCategoriesModule.ts

Co-authored-by: Will Schurman <wschurman@expo.io>

marklawlor committed:
rework NotificationCategoriesModule ib wev

marklawlor commented:
Changes after review:

- rewrote `NotificationCategoriesModule` on web use the types from the `.type` file.
- reverted the changes to the public API in favor of throwing an `UnavailabilityError` on web

marklawlor committed:
build files

marklawlor committed:
Update packages/expo-notifications/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

wschurman commented on packages/expo-notifications/src/NotificationCategoriesModule.ts:
```diff
@@ -0,0 +1,19 @@
+    return [];
+  },
+  async setNotificationCategoryAsync() {
+    throw new UnavailabilityError('Notifications', 'setNotificationCategoryAsync');
```
I think this is okay, but it is technically a somewhat change for web (throws vs returns null). I think it's the correct choice though versus changing the return type to be nullable.
